### PR TITLE
Spinner

### PR DESCRIPTION
**Pull request by MattAWright**
Added Spinners to UI package and a spinner generator function to cmd package. Added spinners to: set, unset, invites accept, invites approve, keypairs generate and keypairs revoke, import, export, view

### CODE REVIEW

File: cmd/invites_approve.go [hunk 1]
```diff
@@ -53,7 +53,10 @@ func invitesApprove(ctx *cli.Context) error {
 		return errs.NewExitError("Invite not found.")
 	}
 
-	err = client.OrgInvites.Approve(context.Background(), *targetInvite, progress)
+	s, p := spinner(fmt.Sprintf("Reviewing invite for %s", email))
```
Comment (ianlivingstone) [inline cmd/invites_approve.go]: I'm thinking the text here should say "Attempting to approve invite for %s" or similar?

File: cmd/import.go [hunk 2]
```diff
@@ -57,7 +58,10 @@ func importCmd(ctx *cli.Context) error {
 		}(secret.value)
 	}
 
-	creds, err := setCredentials(ctx, path, makers)
+	s, p := spinner("Attempting to set credentials...")
```
Comment (ianlivingstone) [inline cmd/import.go]: In some places you're using `...` and others not, we should standardize the messages along with the ones coming back from the daemon.

File: ui/spinner.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+  "github.com/briandowns/spinner"
+)
+
+type Spinner struct{
+  spinner *spinner.Spinner
+  text string
+}
+
+func NewSpinner(text string) *Spinner { return defUI.NewSpinner(text) }
```
Comment (ianlivingstone) [inline ui/spinner.go]: I think this function and the one below should be in the `ui/ui.go` file as thats where the `UI` struct and subsequent global `defaultUI` are defined. Then this file would only contain the spinner struct itself and it's subsequent methods.

File: ui/spinner.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  s.Suffix = " " + text
+  return &Spinner {
+    s,
+    text,
+  }
+}
+
+func (s *Spinner) Start() { defUI.StartSpinner(s) }
+
+func (u *UI) StartSpinner(s *Spinner) {
```
Comment (ianlivingstone) [inline ui/spinner.go]: Why is this method on the `UI` struct inside of the `Spinner` struct? I think it'd be cleaner/clearer lines of responsibility for the `Spinner` struct to have the values it needs off the `UI` struct copied onto itself (or just a reference to the `UI` struct in general).

File: ui/spinner.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
+package ui
+
+import (
+  "time"
+
+  "github.com/briandowns/spinner"
+)
+
+type Spinner struct{
```
Comment (ianlivingstone) [inline ui/spinner.go]: Since these are public methods they'll require documentation strings to pass CI (gofmt)

File: ui/spinner.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+func (u *UI) NewSpinner(text string) *Spinner {
+  s := spinner.New(spinner.CharSets[9], 100*time.Millisecond)
+  s.Suffix = " " + text
+  return &Spinner {
+    s,
+    text,
+  }
+}
+
+func (s *Spinner) Start() { defUI.StartSpinner(s) }
```
Comment (ianlivingstone) [inline ui/spinner.go]: Start/stop should be checking whether or not stdout is attached to a terminal or not. I think, if we're not attached to a terminal, we shouldn't print out the spinners (just make it an empty do-nothing call like several other `ui` functions).

File: cmd/import.go (hunk 2 above)
Comment (MattAWright) [reply cmd/import.go]: Agreed, missed this. Standard is no ellipses

File: ui/spinner.go [hunk 7]
```diff
@@ -0,0 +1,37 @@
+package ui
```
Comment (ianlivingstone) [inline ui/spinner.go]: The indentation in this file is wrong, should be 8 spaces

File: ui/ui.go [hunk 8]
```diff
@@ -81,6 +81,26 @@ type UI struct {
...
+
+func (u *UI) StartSpinner(s *Spinner) {
+  if !u.EnableProgress || !readline.IsTerminal(int(os.Stdout.Fd())) {
+    return
+  }
+
+	s.Start()
+}
+
+func (u *UI) StopSpinner(s *Spinner) {
```
Comment (ianlivingstone) [inline ui/ui.go]: This method should be documented

File: ui/ui.go [hunk 9]
```diff
@@ -81,6 +81,26 @@ type UI struct {
 	EnableColors bool
 }
 
+func (u *UI) NewSpinner(text string) *Spinner {
+	return NewSpinner(text)
+}
+
+func (u *UI) StartSpinner(s *Spinner) {
```
Comment (ianlivingstone) [inline ui/ui.go]: This method should be documented

File: ui/ui.go [hunk 10]
```diff
@@ -81,6 +81,26 @@ type UI struct {
 	EnableColors bool
 }
 
+func (u *UI) NewSpinner(text string) *Spinner {
```
Comment (ianlivingstone) [inline ui/ui.go]: This method should be documented

File: ui/spinner.go (hunk 5 above)
Comment (ianlivingstone) [reply ui/spinner.go]: Also the struct should be documented as well

**Pull request merged**